theano/sandbox/cpu_corr/tests/test_corr.py

```diff
@@ -0,0 +1,242 @@
+        theano_corr = theano.function([input, filters], output, mode=self.mode)
+
+        # initialize input and compute result
+        image_data = numpy.random.random(N_image_shape).astype(self.dtype)
+        filter_data = numpy.random.random(N_filter_shape).astype(self.dtype)
+        try:
+            theano_output = theano_corr(image_data, filter_data)
+        except ValueError:
+            if not should_raise:
+                raise
+            return
+        else:
+            if should_raise:
+                raise Exception("CorrOp should have generated an error")
+
+        # REFERENCE IMPLEMENTATION
+        # Testing correlation, not convolution. Reverse filters.
+        filter_data_corr = numpy.array(filter_data[:, :, ::-1, ::-1],
+                                       copy=True,
+                                       order='C')
+        s = 1.
+        orig_image_data = image_data
+        if border_mode is not 'full':
+            s = -1.
+        out_shape2d = (numpy.array(N_image_shape[-2:]) +
+                       s * numpy.array(N_filter_shape[-2:]) - s)
+        out_shape2d = numpy.ceil(out_shape2d / numpy.array(subsample))
+        out_shape = (N_image_shape[0], N_filter_shape[0]) + tuple(out_shape2d)
+        ref_output = numpy.zeros(out_shape)
+
+        # loop over output feature maps
+        ref_output.fill(0)
+        if border_mode == 'full':
+            image_data2 = numpy.zeros((N_image_shape[0], N_image_shape[1],
+                                       N_image_shape[2] + 2 * N_filter_shape[2] - 2,
+                                       N_image_shape[3] + 2 * N_filter_shape[3] - 2))
+            image_data2[:, :, N_filter_shape[2] - 1:N_filter_shape[2] - 1 + N_image_shape[2],
+                        N_filter_shape[3] - 1:N_filter_shape[3] - 1 + N_image_shape[3]] = image_data
+            image_data = image_data2
+            N_image_shape = image_data.shape
+        for bb in range(N_image_shape[0]):
+            for nn in range(N_filter_shape[0]):
+                for im0 in range(N_image_shape[1]):
+                    filter2d = filter_data_corr[nn, im0, :, :]
+                    image2d = image_data[bb, im0, :, :]
+                    for row in range(ref_output.shape[2]):
+                        irow = row * subsample[0]  # image row
+                        for col in range(ref_output.shape[3]):
+                            icol = col * subsample[1]  # image col
+                            ref_output[bb, nn, row, col] += (image2d[
+                                irow:irow + N_filter_shape[2],
+                                icol:icol + N_filter_shape[3]] * filter2d[::-1, ::-1]
+                            ).sum()
+
+        self.assertTrue(_allclose(theano_output, ref_output))
+
+        # TEST GRADIENT
+        if verify_grad:
+            utt.verify_grad(sym_CpuCorrMM, [orig_image_data, filter_data])
+
+    def test_basic(self):
```

Comment:
I suppose many of those test_* method are just copy/paste from the Conv op test class. If so, why not make this test class inherit from it to remove duplicate code?